File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:

        indices = {}
        pre_tab = [-1] * len(nums2)
        print(pre_tab)

        for idx, num in enumerate(nums2):
            indices[num] = idx

            for j in range(idx + 1, len(nums2)):
                if nums2[j] > num:
                    pre_tab[idx] = nums2[j]
                    break
        output = []
        for num in nums1:
            output.append(pre_tab[indices[num]])
        
        return output
        

File: 0496-next-greater-element-i/0496-next-greater-element-i.cs
public class Solution {
    public int[] NextGreaterElement(int[] nums1, int[] nums2) {
        var stk = new Stack<int>();
        var mp = new Dictionary<int, int>();
        int[] res2 = new int[nums1.Length];

        for (int i = 0; i < nums2.Length; i++) {
            var num = nums2[i];
            while (stk.Count != 0 && stk.Peek() < num) {
                var top = stk.Peek();
                mp[top] = num;
                stk.Pop();
            }
            stk.Push(num);
        }

        int[] res = new int[nums1.Length];
        for(int i = 0; i < nums1.Length; i++) {
            var num = nums1[i];
            if (mp.ContainsKey(num)) res[i] = mp[num];
            else res[i] = -1;
        }
        return res;
    }
}